### PR TITLE
Added AES 128

### PR DESCRIPTION
**Pull request by vshaipov**
This is an AES 128 encryption I added for my pilot project, it unlikely stopped for an economical reason but I'd like to share my work.

### CODE REVIEW

**Comment by vshaipov** [comment]
If my change isn't accepted, please let me know what I need to modify to get accepted it

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,7 @@
 .idea/
 *.iml
 target/
```
Comment (abstractj) [inline .gitignore]: @vshaipov good morning and thank you for the changes. 

Why these changes are required? Eclipse?

File: src/main/java/org/abstractj/kalium/NaCl.java [hunk 2]
```diff
@@ -24,105 +24,130 @@
 
 public class NaCl {
```
Comment (abstractj) [inline src/main/java/org/abstractj/kalium/NaCl.java]: @vshaipov I think these changes here are related with trailing spaces and not relevant to the project, probably caused by the IDE configuration. 

Could you please fix this?

File: src/main/java/org/abstractj/kalium/NaCl.java [hunk 3]
```diff
@@ -24,105 +24,130 @@
...
+				@u_int64_t long length, @In byte[] key);
 
-    /**
-     * This function isn't thread safe. Be sure to call it once, and before performing other operations.
-     *
-     * Check libsodium's <i>sodium_init()</i> documentation for more info.
-     */
-    public static int init() {
-        return sodium().sodium_init();
-    }
```
Comment (abstractj) [inline src/main/java/org/abstractj/kalium/NaCl.java]: Correct me if I'm wrong, but the only changes that really matters to this PR is the AES 128 support, is the code below.
Other than that, the changes above are not relevant, because they are pretty much trailing spaces.

Also please make sure to follow the formatting from the project.

File: src/main/java/org/abstractj/kalium/NaCl.java [hunk 4]
```diff
@@ -24,105 +24,130 @@
...
+		public static final int AES_128_CTR_KEYBYTES = 16;
+		public static final int CRYPTO_STREAM_AES_128_CTR_NONCEBYTES = 16;
+
```
Comment (abstractj) [inline src/main/java/org/abstractj/kalium/NaCl.java]: Include ```public static final``` please, for code consistency.

File: src/main/java/org/abstractj/kalium/crypto/Stream.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+import org.abstractj.kalium.NaCl;
+
+public class Stream {
+
+	private byte[] key;
+
+	public Stream(byte[] key) {
+		this.key = key;
+		checkLength(key, NaCl.Sodium.AES_128_CTR_KEYBYTES);
```
Comment (abstractj) [inline src/main/java/org/abstractj/kalium/crypto/Stream.java]: Make use of import static, please. Into this way we just leave the constant into the code.

File: src/test/java/org/abstractj/kalium/crypto/StreamTest.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
+package org.abstractj.kalium.crypto;
+
+import org.abstractj.kalium.NaCl;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
```
Comment (abstractj) [inline src/test/java/org/abstractj/kalium/crypto/StreamTest.java]: I think we should make use of testing vectors for this. See: https://github.com/abstractj/kalium/blob/master/src/test/java/org/abstractj/kalium/fixture/TestVectors.java

File: .gitignore (hunk 1 above)
Comment (vshaipov) [inline .gitignore]: Yes, it's eclipse. I'll remove it

File: src/main/java/org/abstractj/kalium/NaCl.java (hunk 3 above)
Comment (vshaipov) [inline src/main/java/org/abstractj/kalium/NaCl.java]: Yes, you are right. I corrected trailing spaces

File: src/test/java/org/abstractj/kalium/crypto/StreamTest.java (hunk 6 above)
Comment (vshaipov) [inline src/test/java/org/abstractj/kalium/crypto/StreamTest.java]: I proposed TestVectors.java modification twice. The last one is correct, please ignore the first one

Comment (vshaipov) [inline src/test/java/org/abstractj/kalium/crypto/StreamTest.java]: Please give me rights to modify StreamTest class, I refactored it using constants I added to TetsVectors. See pull request #32

Comment (vshaipov) [inline src/test/java/org/abstractj/kalium/crypto/StreamTest.java]: Here is my code:

package org.abstractj.kalium.crypto;

import static org.abstractj.kalium.encoders.Encoder.HEX;
import static org.abstractj.kalium.fixture.TestVectors.AES128_KEY;
import static org.abstractj.kalium.fixture.TestVectors.AES128_MESSAGE;
import static org.abstractj.kalium.NaCl.Sodium.CRYPTO_STREAM_AES_128_CTR_NONCEBYTES;

import org.junit.Assert;
import org.junit.BeforeClass;
import org.junit.Test;

public class StreamTest {
	private static Stream stream;
	private static byte[] nonce;

	private static byte[] message = AES128_MESSAGE.getBytes();

	@BeforeClass
	public static void init() {
		Random random = new Random();
		stream = new Stream(HEX.decode(AES128_KEY));

		nonce = random
				.randomBytes(CRYPTO_STREAM_AES_128_CTR_NONCEBYTES);
	}

	@Test
	public void testEncryption() {
		byte[] encryptedMessage = stream.encrypt(nonce, message);
		byte[] decryptedMessage = stream.decrypt(nonce, encryptedMessage);
		
		Assert.assertEquals(AES128_MESSAGE, new String(decryptedMessage));
	}
}

**Comment by abstractj** [comment]
@vshaipov we will get there. Will do a second review this week my friend. Thank you for looking at this.

**Comment by vshaipov** [comment]
@abstractj Sure, thank you very much for your time!

**Comment by wsargent** [comment]
Hi -- any progress on this bug?

**Comment by abstractj** [comment]
@wsargent I don't think is a bug, but new feature, correct me if I'm wrong.

Right now, I'm including the testing vectors necessary for this feature. But if this is something you badly need, it can be merged, released like `experimental` until all the testing vectors pass.

Just let me know.

**Comment by vshaipov** [comment]
@wsargent @abstractj it's a feature. Let me know @abstractj if I need to do anything else to have this code included in the project

**Comment by wsargent** [comment]
Hi @abstractj -- now that I look more closely at it, I'm concerned that this is AES-CTR, which isn't authenticated encryption, and not AES-GCM or AES-EAX.  

In keeping with the "boring crypto" theme, can there be a wrapper on top of this method that takes a MAC, or an "unsafe_" prefix or other warning on it?  Something like https://github.com/jedisct1/libsodium/blob/master/src/libsodium/include/sodium/crypto_stream_aes128ctr.h#L5

```
/*
 *  WARNING: This is just a stream cipher. It is NOT authenticated encryption.
 *  While it provides some protection against eavesdropping, it does NOT
 *  provide any security against active attacks.
 *  Unless you know what you're doing, what you are looking for is probably
 *  the crypto_box functions.
 */
```

**Comment by abstractj** [comment]
@wsargent do you think such warning on top of Stream class https://github.com/abstractj/kalium/pull/31/files#diff-afbb21065d0b572424c211b4697e1fe6R12 could help?

**Comment by wsargent** [comment]
@abstractj Yes, but I'd prefer it in the documentation, or even better a class name of `UnauthenticatedStream` that makes it clear this is a stream that doesn't provide authenticated encryption.

**Comment by abstractj** [comment]
@wsargent yes, at the same time Kalium is just a wrapper for libsodium. So I think that a comment on top of the Stream class will be enough, people using libsodium are already aware of it, that's my hope.

At this moment I'm not planning to rush in this, but if somehow you would like to see this change in, let me know.

**Comment by wsargent** [comment]
@abstractj So I see a number of problems with that.

The first is that I don't expect people who use Kalium to be aware of libsodium -- they're not necessarily going to look at a different library, even one that it wraps.

The second is that I don't think the people who use libsodium are aware of the issue either.  If you look at the documentation page, http://doc.libsodium.org/advanced/aes-128-ctr.html -- there are notes at the bottom of the page, but it's not in bold font.  The warning it provides talks in technical terms about an "authentication tag", but this assumes that people understand what that means and the practical implications behind the lack of authentications:

> crypto_stream_aes128ctr() doesn't have an internal counter: the nonce itself is incremented after each 16-bytes block. As a result, extreme care should be taken to avoid reusing a nonce with the same key.
> Unless using AES is a requirement, you should consider crypto_stream (XSalsa20) or crypto_stream_chacha20 (ChaCha20) instead.
> crypto_stream_aes128ctr_xor() simply combines the message with the output of the stream cipher with the XOR operation, and does not add any authentication tag.
> Unless this is exactly what you need, you should use crypto_secretbox() or crypto_box() instead.

This is not nearly as clear as the warning inside the C header itself, because it doesn't say that it doesn't provide any security against active attacks.  The manual does put AES-128-CTR under the "advanced" section, and the "Stream Ciphers" [text](http://doc.libsodium.org/advanced/stream_ciphers.html) contains 

> These functions are stream ciphers. They do not provide authenticated encryption.

but that STILL doesn't go the extra step to explain that unauthenticated encryption is insecure.

The third issue (and I admit having worked as a consultant I am somewhat biased) is that if there's a way to do something wrong, by $DIETY programmers will find a way to do it -- and NaCl / libsodium is a reaction against that, with settings that are almost impossible to screw up.  Except that... well, now programmers are even less likely to RTFM, because they "know" that anything on top of libsodium won't let them do it wrong.  Except in this case, it will.

To your final point, I don't need AES for anything I'm doing right now -- I'm looking at both Kalium and Keyczar and looking to see if it's possible to put an abstract crypto layer on top.

**Comment by abstractj** [comment]
@wsargent one way or another people must be aware of libsodium, otherwise there's no way to use Kalium. But I got your point.

Thanks a lot for the valuable input, I will take this into consideration before merge this pull-request and will think about how to properly provide some guidance for programmers.

**Comment by bendoerr** [comment]
Just a thought. AES128-CTR is no longer included libsodium's documentation and if I recall is not hardware accelerated and only there for backwards compatibility with NaCl. Also according to [https://github.com/jedisct1/libsodium/issues/317#issuecomment-154719579] may not even be available by default in the future.

**Comment by abstractj** [comment]
@bendoerr thank you for the heads up. @vshaipov I think we have more cons than pros here, I'm closing this change for now and I really appreciate your contribution.
